branchLayoutImpl/src/test/java/com/virtuslab/branchlayout/BranchLayoutFileReaderTest.java

```diff
@@ -0,0 +1,122 @@
+    PowerMockito.when(BranchLayoutFileUtils.getIndentWidth(anyString(), anyChar())).thenCallRealMethod();
+    return reader;
+  }
+
+  @Test
+  public void read_givenCorrectFile_reads() throws Exception {
+    // given
+    List<String> linesToReturn = List.of("A", " B", "C");
+    BranchLayoutFileReader reader = getBranchLayoutFileReaderForLines(linesToReturn, /* indentWidth */ 1);
+
+    // when
+    BranchLayout branchLayout = reader.read();
+
+    // then
+    Assert.assertTrue(branchLayout.findEntryByName("A").isDefined());
+    Assert.assertTrue(branchLayout.findEntryByName("B").isDefined());
+    Assert.assertTrue(branchLayout.findEntryByName("C").isDefined());
+    Assert.assertEquals(2, branchLayout.getRootBranches().size());
+  }
+
+  @Test
+  public void read_givenCorrectFileWithRootsOnly_reads() throws Exception {
+    // given
+    List<String> linesToReturn = List.of("A", "B");
+    BranchLayoutFileReader reader = getBranchLayoutFileReaderForLines(linesToReturn, /* indentWidth */ 1);
+
+    // when
+    BranchLayout branchLayout = reader.read();
+
+    // then
+    Assert.assertTrue(branchLayout.findEntryByName("A").isDefined());
+    Assert.assertTrue(branchLayout.findEntryByName("B").isDefined());
+    Assert.assertEquals(2, branchLayout.getRootBranches().size());
+  }
+
+  @Test
+  public void read_givenEmptyFile_reads() throws Exception {
+    // given
+    List<String> linesToReturn = List.empty();
+    BranchLayoutFileReader reader = getBranchLayoutFileReaderForLines(linesToReturn, /* indentWidth */ 1);
+
+    // when
+    BranchLayout branchLayout = reader.read();
+
+    // then no exception thrown
+    Assert.assertEquals(0, branchLayout.getRootBranches().size());
+  }
+
+  @Test
+  public void read_givenFileWithIndentedFirstEntry_throwsException() throws Exception {
+    // given
+    List<String> linesToReturn = List.of(" A", " B");
+    BranchLayoutFileReader reader = getBranchLayoutFileReaderForLines(linesToReturn, /* indentWidth */ 1);
+
+    // when
+    BranchLayoutException exception = assertThrows(BranchLayoutException.class, () -> reader.read());
+
+    // then
+    int i = exception.getErrorLine().get();
+    assertEquals(1, i);
+  }
+
+  @Test
+  public void read_givenFileWithIndentWidthNotAMultiplicityOfLevelWidth_throwsException() throws Exception {
+    // given
+    List<String> linesToReturn = List.of("A", "   B", " C");
+    BranchLayoutFileReader reader = getBranchLayoutFileReaderForLines(linesToReturn, /* indentWidth */ 3);
+
+    // when
+    BranchLayoutException exception = assertThrows(BranchLayoutException.class, () -> reader.read());
+
+    // then
+    int i = exception.getErrorLine().get();
+    assertEquals(3, i);
+  }
+
+  @Test
+  public void read_givenFileWithSubentryIndentGreaterThanOneToParent_throwsException() throws Exception {
+    // given
+    List<String> linesToReturn = List.of("A", "  B", "      C");
```

Comment:
Add some empty lines so that the significant line number != overall line number

Reply:
In this and/or other test cases ofc

Reply:
Good point! :1st_place_medal: Actually, the exception is thrown when the first entry is intended